### PR TITLE
Manage members in organization [INS-2705]

### PR DESCRIPTION
**Pull request by yaoweiprc**
Now users can 
- invite new members
- change member's role
- remove existing member
- revoke invitation
- search members and invites
in app. It's basically migrated from insomnia-website-app. The logic for permission validation and error messages is essentially the same as in the web version. Except that:
- existing members and pending invites are combined into one list
- users can choose role when inviting
image

Other changes:
- Add an react hook `useStateRef` to get latest state immediately after setState, rather than in next round of render.
- Add an extra param `onlyResolveOnSuccess` to insomniaFetch so that it can reject when the response status code is not 2xx.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "*.njson": "ndjson"
   },
   "files.insertFinalNewline": true,
-  "editor.formatOnSave": true,
+  "editor.formatOnSave": false,
```
Comment (jackkav) [inline .vscode/settings.json]: please keep dx changes in a separate PR, add a PR description, and mention anything you could have done but didn't want to include in the scope. For more clarity do the some PRs reviews over zoom with the team.

Comment (yaoweiprc) [reply .vscode/settings.json]: Current formatter use 4 spaces as tab on my computer. I did not find where to set it. So I just disabled formatOnSave temporarily. 
I have changed it back. I'll try to use some other formatter like Prettier in a separate PR when I'm free.

**Comment by cwangsmv** [comment]
It is better to show a skeleton like web when user opens the invite modal. Users will see "no member" first for now.

https://github.com/user-attachments/assets/128e3c67-6de0-4a21-bcc8-818d4c79e97f

**Comment by cwangsmv** [comment]
Should we change the title and hide stack trace for RBAC error:
Screenshot 2024-09-20 at 14 33 42

Web behavior:

Screenshot 2024-09-20 at 14 35 09

File: .vscode/settings.json [hunk 2]
```diff
@@ -15,4 +15,9 @@
   "[jsonc]": {
     "editor.defaultFormatter": "vscode.json-language-features"
   },
+  "cSpell.words": [
```
Comment (CurryYangxx) [inline .vscode/settings.json:18]: What are these configurations for?

File: packages/insomnia/src/ui/routes/organization.tsx [hunk 3]
```diff
@@ -645,16 +647,19 @@ const OrganizationRoute = () => {
...
                     <span className="truncate">
                       Invite
                     </span>
                   </Button>
+                  <InviteModalContainer
+                    {...{
+                      isOpen: isInviteModalOpen,
+                      setIsOpen: setIsInviteModalOpen,
+                    }}
+                  />
```
Comment (CurryYangxx) [inline packages/insomnia/src/ui/routes/organization.tsx:662]: Why we need destruction here?

**Comment by CurryYangxx** [comment]
This pr include some changes that are not part of this feature:

- add a new react hook
- add a new param  in insomniaFetch

It is better to explain these changes in pr description to make review easier

File: packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+        let json;
+        try {
+          json = await error.response.json();
+        } catch (e) {
+          throw new Error(`Failed to invite ${inviteeEmail}`);
+        }
+        if (json?.message) {
+          throw new Error(json.message);
+        }
+      }
```
Comment (CurryYangxx) [inline packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx]: I think we can do these logic in insomniaFetch function.

File: packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+you don’t have more than one project. Since you have more than one project, you need to upgrade to
+Individual or above to continue.`
+        );
+      }
+      if (json?.error === NEEDS_TO_INCREASE_SEATS_ERROR) {
+        throw new Error(needToIncreaseSeatErrMsg);
+      }
+      if (json?.message) {
+        throw new Error(json.message);
+      }
```
Comment (CurryYangxx) [inline packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx:260]: same thing

File: packages/insomnia/src/ui/insomniaFetch.ts [hunk 6]
```diff
@@ -40,6 +59,12 @@ export async function insomniaFetch<T = void>({ method, path, data, sessionId, o
     if (uri) {
       window.main.openDeepLink(uri);
     }
+    if (onlyResolveOnSuccess && !response.ok) {
+      throw new ResponseFailError(
+        `${response.status} ${response.statusText}`,
+        response,
+      );
+    }
```
Comment (CurryYangxx) [inline packages/insomnia/src/ui/insomniaFetch.ts]:
```suggestion
    if (onlyResolveOnSuccess && !response.ok) {
      const res = await response.json();
      throw new ResponseFailError(
        `res.message`,
      );
    }
```

File: packages/insomnia/src/ui/components/modals/invite-modal/invite-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,671 @@
...
+    return 0;
+  }
+  const validMaxPage = Math.floor(total / PAGE_SIZE);
+  if (page > validMaxPage) {
+    return validMaxPage;
+  }
+  return page;
+}
+
+const usePagination = (permissionRef: MutableRefObject<Record<Permission, boolean>>) => {
```
Comment (CurryYangxx) [inline packages/insomnia/src/ui/components/modals/invite-modal/invite-modal.tsx:459]: This seems like a hook, should we need to move it to hooks dir?

**Comment by CurryYangxx** [comment]
There are much typescript definition and util functions in component file, we can discuss if it is better to move them to some other dir together.

File: packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+export const InviteForm = ({ onInviteCompleted, allRoles }: {
+  onInviteCompleted?: () => void;
+  allRoles: Role[];
+}) => {
+  const selectedRoleRef = React.useRef<Role>(
+    allRoles.find(role => role.name === defaultRoleName) as Role,
+  );
+  const organizationId = useParams().organizationId as string;
+  const [inputStr, setInputStr] = useState('');
+  const [addedMembers, setAddedMembers, addedMembersRef] = useStateRef<string[]>([]);
```
Comment (CurryYangxx) [inline packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx:24]: why we use this `useStateRef` hook here?

File: .vscode/settings.json (hunk 2 above)
Comment (yaoweiprc) [reply .vscode/settings.json:18]: For spell check.

File: packages/insomnia/src/ui/routes/organization.tsx (hunk 3 above)
Comment (yaoweiprc) [reply packages/insomnia/src/ui/routes/organization.tsx:662]: To pass props to components. I prefer this style.

File: packages/insomnia/src/ui/components/modals/invite-modal/invite-modal.tsx (hunk 7 above)
Comment (yaoweiprc) [reply packages/insomnia/src/ui/components/modals/invite-modal/invite-modal.tsx:459]: This is a hook specifically used here for pagination and is not reusable.

File: packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx (hunk 8 above)
Comment (yaoweiprc) [reply packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx:24]: To get latest state immediately after setState, rather than in next round of render.

File: packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx (hunk 4 above)
Comment (yaoweiprc) [reply packages/insomnia/src/ui/components/modals/invite-modal/organization-invite-form.tsx]: I've talked about it with Jack before. Previously, when the status code of a request response was not 2xx, we were unable to retrieve the status code information from the insomniaFetch return. When the status code is not 2xx, the response may contain different information. As a general request method, insomniaFetch should not handle specific failure cases but should return the failed response to the caller, allowing the caller to handle the failure according to the specific API. This modification enables minimal intrusion to insomniaFetch while maintaining compatibility.

File: packages/insomnia/src/ui/insomniaFetch.ts (hunk 6 above)
Comment (yaoweiprc) [reply packages/insomnia/src/ui/insomniaFetch.ts]: When response.ok is false, the content-type of the response is not necessarily JSON, and the JSON structure may not always contain a message property. I believe we should return the response directly to the caller and let the caller decide how to handle the error.

**Comment by yaoweiprc** [comment]
> This pr include some changes that are not part of this feature:
> 
> * add a new react hook
> * add a new param  in insomniaFetch
> 
> It is better to explain these changes in pr description to make review easier

Yeah, I have added it.

**Comment by notjaywu** [comment]
> Should we change the title and hide stack trace for RBAC error: Screenshot 2024-09-20 at 14 33 42
> 
> Web behavior:
> 
> Screenshot 2024-09-20 at 14 35 09

image

**Comment by yaoweiprc** [comment]
> It is better to show a skeleton like web when user opens the invite modal. Users will see "no member" first for now.
> 
>  Screen.Recording.2024-09-20.at.14.23.57.mp4

image
Add a spin icon when loading.

**Comment by yaoweiprc** [comment]
> Should we change the title and hide stack trace for RBAC error: Screenshot 2024-09-20 at 14 33 42
> 
> Web behavior:
> 
> Screenshot 2024-09-20 at 14 35 09

image
Use another alert modal.

File: packages/insomnia/src/ui/insomniaFetch.ts (hunk 6 above)
Comment (yaoweiprc) [reply packages/insomnia/src/ui/insomniaFetch.ts]: Solved in https://github.com/Kong/insomnia/pull/7949/commits/e98eb1df8bd0b7db48846ca4e2c63f5e8a766a4a

**Review by cwangsmv — DISMISSED** [review]
lgtm

**Review by cwangsmv — APPROVED** [review]
